src/geopackage/read.cpp

```diff
@@ -0,0 +1,108 @@
+
+    // Check if layer exists
+    if (!db.has_table(layer)) {
+        // Since the layer doesn't exist, we need to output some additional
+        // debug information with the error. In this case, we add ALL the tables
+        // available in the GPKG, so that if the user sees this error, then it
+        // might've been either a typo or a bad data input, and they can correct.
+        std::string errmsg = "[" + std::string(sqlite3_errmsg(db.connection())) + "] " +
+                             "table " + layer + " does not exist.\n\tTables: ";
+
+        auto errquery = db.query("SELECT name FROM sqlite_master WHERE type='table'").next();
+        while(!errquery.done()) {
+            errmsg += errquery.get<std::string>(0);
+            errmsg += ", ";
+            errquery.next();
+        }
+
+        throw std::runtime_error(errmsg);
+    }
+
+    // Layer exists, getting statement for it
+    //
+    // this creates a string in the form:
+    //     WHERE id IN (?, ?, ?, ...)
+    // so that it can be bound by SQLite.
+    // This is safer than trying to concatenate
+    // the IDs together.
+    std::string joined_ids = "";
+    if (!ids.empty()) {
+        joined_ids = " WHERE id IN (?";
+        for (size_t i = 1; i < ids.size(); i++) {
+            joined_ids += ", ?";
+        }
+        joined_ids += ")";
+    }
+
+    // Get number of features
+    sqlite_iter query_get_layer_count = db.query("SELECT COUNT(*) FROM " + layer + joined_ids, ids);
+    query_get_layer_count.next();
+    const int layer_feature_count = query_get_layer_count.get<int>(0);
+
+    #ifndef NGEN_QUIET
+    // output debug info on what is read exactly
+    std::cout << "Reading " << layer_feature_count << " features in layer " << layer;
+    if (!ids.empty()) {
+        std::cout << " (id subset:";
+        for (auto& id : ids) {
+            std::cout << " " << id;
+        }
+        std::cout << ")";
+    }
+    std::cout << std::endl;
+    #endif
+
+    // Get layer feature metadata (geometry column name + type)
+    sqlite_iter query_get_layer_geom_meta = db.query("SELECT column_name FROM gpkg_geometry_columns WHERE table_name = ?", layer);
+    query_get_layer_geom_meta.next();
+    const std::string layer_geometry_column = query_get_layer_geom_meta.get<std::string>(0);
+
+    // Get layer
+    sqlite_iter query_get_layer = db.query("SELECT * FROM " + layer + joined_ids, ids);
+    query_get_layer.next();
+
+    // build features out of layer query
+    std::vector<geojson::Feature> features;
+    features.reserve(layer_feature_count);
+    while(!query_get_layer.done()) {
+        geojson::Feature feature = build_feature(
+            query_get_layer,
+            layer_geometry_column
+        );
+
+        features.push_back(feature);
+        query_get_layer.next();
+    }
+
+    // get layer bounding box from features
+    //
+    // GeoPackage contains a bounding box in the SQLite DB,
+    // however, it is in the SRS of the GPKG. By creating
```

Comment:
I couldn't figure out from the Boost::Geometry documentation - what's SRS stand for? It only seems to appear as `namespace boost::geometry::srs` which contains `class spheroid`

Reply:
SRS stands for **Spatial Reference System**, it's the same thing as CRS, which is **Coordinate Reference System**. Here's a set of slides that's useful for showing boost's use of srs transformations: [SRS Transformations](https://archive.fosdem.org/2019/schedule/event/geo_spatialboostgeometry/attachments/slides/3306/export/events/attachments/geo_spatialboostgeometry/slides/3306/Boost_Geometry_SRS_Transformations.pdf).

Unfortunately (in my opinion), boost's namespacing of the SRS functions is really poor. The primary one that you'll need to be concerned with (as far as this PR) is `boost::geometry::srs::transformation`, which is defined in `<boost/geometry/srs/transformation.hpp>`.